### DISCUSSION TITLE
delete selected records

### DISCUSSION THREAD

**Discussion by samaremad** [serenity-is/Serenity#5448] [Help] [answered]
hello everyone

i want to add checkboxes to the table so that i can select more than one record and press on button to delete them how can i do that ?

**Comment by minhhungit** [comment, marked as answer]
there is a sample to create checkbox here https://serenity.is/demo/BasicSamples/CancellableBulkAction

code https://github.com/serenity-is/Serene/blob/master/Serene/Serene.Core/Modules/BasicSamples/Grids/CancellableBulkAction/CancellableBulkActionGrid.ts

after you have checkbox, just add new button and call delete service in click() method

**Reply by samaremad** [reply]
first thank you for your reply 
this is exactly what i want to do but the problem that i already created the function getbuttons so how can i edit it without getting so many errors 

 getButtons() { 
            
            var buttons = super.getButtons();
            var MyRow = NomineeRow.Fields;

            buttons.push(KRVoting.Common.ExcelExportHelper.createToolButton({
                grid: this,
                onViewSubmit: () => this.onViewSubmit(),
                service: NomineeService.baseUrl + '/ListExcel',
                separator: true,
               // hint: "This is the Hint",
                title: "Nominee Report",
               
            }));

            return buttons;
        }

**Reply by samaremad** [reply]
hello mr. jin 

in OrderBulkAction()
what i must write to delete the selected records ?

**Reply by minhhungit** [reply]
If you create toolbar button, in the button's click event, you need to get row Id to delete, then call service to delete it

To get ID of selected row, you can use rowSelection mixin, you can find example in [CancellableBulkActionGrid](https://serenity.is/demo/BasicSamples/CancellableBulkAction)

```javascript
this.rowSelection = new Serenity.GridRowSelectionMixin(this);
```

Here is full code I wrote

```javascript
namespace SerenitySkins.Northwind {

    @Serenity.Decorators.registerClass()
    export class CategoryGrid extends Serenity.EntityGrid<CategoryRow, any> {
        protected getColumnsKey() { return "Northwind.Category"; }
        protected getDialogType() { return <any>CategoryDialog; }
        protected getIdProperty() { return CategoryRow.idProperty; }
        protected getLocalTextPrefix() { return CategoryRow.localTextPrefix; }
        protected getService() { return CategoryService.baseUrl; }

        private rowSelection: Serenity.GridRowSelectionMixin;

        constructor(container: JQuery) {
            super(container);
        }

        protected createToolbarExtensions() {
            super.createToolbarExtensions();
            this.rowSelection = new Serenity.GridRowSelectionMixin(this);
        }

        protected getButtons() {
            let button = super.getButtons();

            button.push({
                title: "Hit me to delete record",
                cssClass: "text-red",
                icon: "fa-times text-red",
                onClick: () => {
                    let selectedKeys = this.rowSelection.getSelectedKeys();
                    if (selectedKeys != null && typeof selectedKeys != 'undefined' && selectedKeys.length > 0) {
                        if (selectedKeys.length > 1) {
                            Q.warning("Should select only one row");
                        }
                        else {
                            CategoryService.Delete({ EntityId: Q.toId(selectedKeys[0]) }, res => {
                                Q.notifySuccess("Row is deleted");
                                this.refresh();
                            });
                        }
                    }
                    else {
                        Q.warning("Please select one row to delete");
                    }
                }
            })

            return button;
        }

        protected getColumns() {
            var columns = super.getColumns();
            columns.splice(0, 0, Serenity.GridRowSelectionMixin.createSelectColumn(() => this.rowSelection));
            return columns;
        }
    }
}
```

**Reply by minhhungit** [reply]
In case you add inline button on row then [check this](https://github.com/serenity-is/Serenity/discussions/5411#discussioncomment-321300)

**Reply by samaremad** [reply]
thank you so much it is work correctly but this condition didnt work
else {
                        Q.warning("Please select one row to delete");
                    }

when i press button without select records it goes to delete procedure and give me error becouse the id didnt match in database how can i solve that
i try to get breakpoint but it dosent work

**Reply by minhhungit** [reply]
Add more condition: **&& selectedKeys.length > 0**

```javascript
onClick: () => {
    let selectedKeys = this.rowSelection.getSelectedKeys();
    if (selectedKeys != null && typeof selectedKeys != 'undefined' && selectedKeys.length > 0) {
        if (selectedKeys.length > 1) {
            Q.warning("Should select only one row");
        }
        else {
            CategoryService.Delete({ EntityId: Q.toId(selectedKeys[0]) }, res => {
                Q.notifySuccess("Row is deleted");
                this.refresh();
            });
        }
    }
    else {
        Q.warning("Please select one row to delete");
    }
}
```

**Reply by samaremad** [reply]
i did that but i get alert not the warning message
![Screenshot (70)](https://user-images.githubusercontent.com/77009503/107485200-21b47480-6b8c-11eb-868b-69f167642177.png)

**Reply by minhhungit** [reply]
debug it please, the error is your code, the error showed that one of your variables is null
it looks you wrote something in your repository or endpoint

**Reply by samaremad** [reply]
i want to debug the TS. file do you know how to do that?

**Reply by samaremad** [reply]
hi my friend how are you ?
this function to delete just 1 record if use 


```
for (var i = 0; i < selectedKeys.length; i++)
                            NomineeService.Delete({ EntityId: Q.toId(selectedKeys[i]) }, res => {
                                
                               this.refresh();
                           });

```


instead of 

`Q.warning("Should select only one row");`


it dosent work correctly and get to deadlock

**Reply by minhhungit** [reply]
Sorry but your code is not good
Just use my code when you need to delete only one item

If you want to delete multiple item, should remove **this.refresh()** out of for loop 
**But** it's not a good way too, better way you will create new endpoint and pass all Ids instead delete one by one, then on server side, you will use for loop to delete record. that means you just call only one request to server and it will include all Ids, after that you can call **this.refresh()**

**Reply by luongnv6** [reply]
Create a class:
```C#
public class DeleteMultiRequest: ServiceRequest
    {
        public List<int> Ids { get; set; }
    }
```
in EndPoint create a new method:
```C#
 [HttpPost, AuthorizeDelete(typeof(MyRow))]
        public ServiceResponse DeleteMulti(IUnitOfWork uow, DeleteMultiRequest request)
        {
            request.Ids.ForEach(v =>
            {
                new MyRepository().Delete(uow, new DeleteRequest() { EntityId = v });
            });

            return new ServiceResponse();
        }
```
event click ôn button delete:
```C#
let selectedKeys = this.rowSelection.getSelectedKeys();
    if (selectedKeys != null && typeof selectedKeys != 'undefined' && selectedKeys.length > 0) {
            CategoryService.DeleteMulti({ Ids: selectedKeys.getSelectedAsInt64() }, res => {
                Q.notifySuccess("Row is deleted");
                this.refresh();
            });
    }
    else {
        Q.warning("Please select one row to delete");
    }
